SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,647 @@
+
+	if (Entities.Num() == 0)
+	{
+		return;
+	}
+	if (Regions.Num() == 0)
+	{
+		return;
+	}
+
+	// Set an entity that has not moved since 5 secs as inactive. It will not be checked against the regions.
+	// This assumes that the bits allocation against workers remains constant, which is not valid anymore when changing/reassigning regions.
+	const double InactiveTime = 5.0;
+	const uint64 TimeToConsiderInactive = FPlatformTime::Cycles64() - (InactiveTime / FPlatformTime::GetSecondsPerCycle64());
+	for (int32 i = 0; i < (int32)ActiveEntities; ++i)
+	{
+		if (ActiveTimestamp[i] < TimeToConsiderInactive)
+		{
+			SwapEntries(i, ActiveEntities - 1);
+			--ActiveEntities;
+			--i;
+		}
+	}
+
+	check(Regions.Num() <= 64);
+
+	const uint32 NumRegions = Regions.Num();
+	const uint32 NumEntities = Entities.Num();
+
+	if (CachedServerInterest[0].Num() < (int32)NumRegions)
+	{
+		CachedServerInterest[0].SetNum(NumRegions);
+		CachedServerInterest[1].SetNum(NumRegions);
+		CachedServerInterest[2].SetNum(NumRegions);
+	}
+
+	constexpr bool bUseBroadphaseForRegions = false;
+
+	if (Broadphase && bUseBroadphaseForRegions)
+	{
+		// 60ms, 200k entities, 64 regions, 5% overlap
+		// Worse than the non-spatialized version, which is not surprising, since we are visiting all the entities anyway,
+		// hence visiting the tree before testing the entities is just more data to process.
+		// Have to test it for the player interest since the tree culling might actually do something in that case.
+
+		// Since the regions are a partition of the entire space in the server case, a better strategy would be to find the region in the
+		// partition an entity belongs to and test the position against the segment boundaries with the interest border. The region's
+		// connectivity would be needed to interpret these results.
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBroadphase);
+		const uint32 NumActiveEntities = ActiveEntities;
+
+		uint64* VisibilityPtr = Visibility.GetData();
+		const uint32* Flags = EntityFlags.GetData();
+		for (uint32 i = 0; i < NumActiveEntities; ++i)
+		{
+			const uint64 Mask = UINT64_MAX;
+			*VisibilityPtr = ((uint64)(*Flags != 0)) * Mask;
+			++VisibilityPtr;
+			++Flags;
+		}
+
+		Broadphase->ComputeBoxVisibility(Regions, Visibility.GetData());
+	}
+
+	constexpr bool bVectorizeRegions = true;
+	constexpr bool bVectorizeEntities = false;
+
+	if (!bVectorizeRegions && !bVectorizeEntities)
+	{
+		// Not-so naively iterate over the region
+		// 24ms, 200k entities, 64 regions
+		SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputationBox);
+		const FVector2D* Pos = EntityPosition.GetData();
+		const uint32* Flags = EntityFlags.GetData();
+		uint64* VisibilityPtr = Visibility.GetData();
+
+		const uint32 NumActiveEntities = ActiveEntities;
+
+		for (uint32 i = 0; i < NumActiveEntities; ++i)
+		{
+			const uint64 AllRegionsMask = UINT64_MAX;
+			*VisibilityPtr = ((uint64)(*Flags != 0)) * AllRegionsMask;
```

Comment:
worth wrapping in a descriptive inline free function since we do it more than once?